packages/opentelemetry-scope-zone-peer-dep/src/ZoneScopeManager.ts

```diff
@@ -0,0 +1,511 @@
+    this._saveOriginalRemoveEventWithScope(scope, target, original);
+    return function(this: {}, event: string, listener: Func<void>) {
+      if (
+        target.__ot_listeners === undefined ||
+        target.__ot_listeners[event] === undefined
+      ) {
+        return original.call(this, event, listener);
+      }
+      const events = target.__ot_listeners[event];
+      const patchedListener = events.get(listener);
+      scopeManager._cleanEventTargetsFromScope(scope, target);
+      return original.call(this, event, patchedListener || listener);
+    };
+  }
+
+  /**
+   * Restores previous scope for certain zone
+   * @param zone
+   */
+  private _restorePreviousScope(zone: Zone) {
+    const zoneName = zone.name;
+    const scopes = this._scopes[zoneName];
+    if (scopes && scopes.length > 0) {
+      this._cleanEventTargetsFromScope(scopes[scopes.length - 1]);
+      scopes.splice(scopes.length - 1, 1);
+    }
+  }
+
+  /**
+   * Saves original addEventListener method so it can be restored later
+   * @param scope
+   * @param target
+   * @param original original addEventListener method
+   */
+  private _saveOriginalAddEventWithScope(
+    scope: unknown,
+    target: TargetWithEvents,
+    original: any
+  ) {
+    const foundScope = this._ensureScopeWithEventTargetsExists(scope, target);
+    if (foundScope) {
+      foundScope.eventTargets.addEventListener = original;
+    }
+  }
+
+  /**
+   * Saves original removeEventListener method so it can be restored later
+   * @param scope
+   * @param target
+   * @param original original removeEventListener method
+   */
+  private _saveOriginalRemoveEventWithScope(
+    scope: unknown,
+    target: TargetWithEvents,
+    original: any
+  ) {
+    const foundScope = this._ensureScopeWithEventTargetsExists(scope, target);
+    if (foundScope) {
+      foundScope.eventTargets.removeEventListener = original;
+    }
+  }
+
+  /**
+   * Returns the active scope
+   */
+  active(): unknown | undefined {
+    const activeZoneName = this._getActiveZoneName();
+
+    const active = this._activeScopeFromZoneName(activeZoneName);
+    if (active) {
+      return active;
+    }
+    if (this._enabled) {
+      return window;
+    }
+    return undefined;
+  }
+
+  /**
+   * Binds a the certain scope or the active one to the target function and then returns the target
+   * @param target
+   * @param scope
+   */
+  bind<T>(target: T | TargetWithEvents, scope?: unknown): T {
+    // if no specific scope to propagate is given, we use the current one
+    if (scope === undefined) {
+      scope = this.active();
+    }
+    if (typeof target === 'function') {
+      return this._bindFunction(target, scope);
+    } else if (isListenerObject(target)) {
+      this._bindListener(target, scope);
+    }
+    return (target as unknown) as T;
+  }
+
+  /**
+   * Disable the scope manager (clears all the scopes)
+   */
+  disable(): this {
+    const zoneNames = Object.keys(this._scopes);
```

Comment:
Maybe, `Object.keys(this._scopes).forEach(zoneName => this._cleanScopesFromZone(zoneName));`